calibrate-touchscreen/src/main.rs

```diff
@@ -0,0 +1,444 @@
+use std::{fs::File, os::unix::prelude::AsFd};
+use std::cmp::max;
+use std::collections::{HashMap};
+use std::error::Error;
+
+use wayland_client::{delegate_noop, protocol::{
+    wl_buffer, wl_keyboard, wl_registry, wl_seat, wl_shm, wl_shm_pool, wl_output,
+    wl_touch, wl_pointer
+}, Connection, Dispatch, QueueHandle, WEnum, Proxy};
+use wayland_client::protocol::wl_compositor::WlCompositor;
+use wayland_client::protocol::wl_output::WlOutput;
+use wayland_client::protocol::wl_shm::WlShm;
+use wayland_client::protocol::wl_surface::WlSurface;
+
+use wayland_protocols::xdg::shell::client::{xdg_surface, xdg_toplevel, xdg_wm_base};
+use wayland_protocols::xdg::shell::client::xdg_wm_base::XdgWmBase;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let conn = Connection::connect_to_env()?;
+
+    let mut event_queue = conn.new_event_queue();
+    let qhandle = event_queue.handle();
+
+    let display = conn.display();
+    display.get_registry(&qhandle, ());
+
+    let mut state = State {
+        running: true,
+        compositor: None,
+        wm_base: None,
+        wl_shm: None,
+        initialised: None,
+        pointer_x: 0f64,
+        pointer_y: 0f64,
+        pointer_width: 0f64,
+        pointer_height: 0f64,
+    };
+
+    println!("Starting the calibrate-touchscreen app: touch the target spots.");
+    println!("(Or press <ESC> to quit!)");
+
+    while state.not_done() {
+        event_queue.blocking_dispatch(&mut state)?;
+    }
+
+    Ok(())
+}
+
+struct FullscreenSurface {
+    width: i32,
+    height: i32,
+    wl_surface: WlSurface,
+}
+
+impl FullscreenSurface {
+    fn attach_buffer(&self, wl_shm: &WlShm, qh: &QueueHandle<State>, targets_index: usize) {
+        let width = self.width;
+        let height = self.height;
+
+        let mut file = tempfile::tempfile().unwrap();
```

Comment:
Ideally this would be a memfd (or at least a shm fd) rather than a file-backed thing, but it doesn't really matter.